src/main.c

```diff
@@ -142,10 +148,131 @@ static bool is_valid_stream(char *s)
 	return true;
 }
 
+char* get_saved_file_path(void)
+{
+	char* home_path = getenv("XDG_STATE_HOME");
+	char* fallback_path = NULL;
+
+	if (home_path == NULL)
+	{
+		// fallback to $HOME/.local/state
+		home_path = getenv("HOME");
+
+		if (home_path == NULL)
+			return NULL;
+
+		const char* local_path = "/.local/state";
+		size_t len_home_path = strlen(home_path);
+		size_t len_local_path = strlen(local_path);
+
+		fallback_path = malloc(len_home_path + len_local_path + 1);
+
+		strcpy(fallback_path, home_path);
+		strcat(fallback_path, local_path);
+
+		home_path = fallback_path;
+	}
+
+	size_t len_home_path = strlen(home_path);
+
+	const char* dir_name = "/nudoku/";
+	size_t len_dir_name = strlen(dir_name);
+
+	char* dir_path = malloc(len_home_path + len_dir_name + 1);
+	strcpy(dir_path, home_path);
+	strcat(dir_path, dir_name);
+
+	struct stat st = {0};
+	if (stat(dir_path, &st) == -1)
+		mkdir(dir_path, 0700);
+
+	const char* file_name = STATE_FILE_NAME;
+	size_t len_file = strlen(file_name);
+	size_t len_dir_path = strlen(dir_path);
+
+	char* file_path = malloc(len_dir_path + len_file + 1);
+	if (file_path == NULL)
+	{
+		free(dir_path);
+		free(fallback_path);
+		return NULL;
+	}
+
+	strcpy(file_path, dir_path);
+	strcat(file_path, file_name);
+
+	free(dir_path);
+	free(fallback_path);
+
+	return file_path;
+}
+
+bool get_board_save(char user_board[], char plain_board[])
+{
+	int c;
+	int i = 0;
+	char board[2 * STREAM_LENGTH];
+	char tmp_board[STREAM_LENGTH];
+
+	char* file_path = get_saved_file_path();
+	FILE* fp = fopen(file_path, "r");
+	if ( fp == NULL )
+		return false;
+
+	while ( (c = fgetc(fp)) != EOF )
+	{
+		board[i++] = c;
+	}
+
+	fclose(fp);
+	unlink(file_path);
+	free(file_path);
+
+	strncpy(user_board, board, STREAM_LENGTH);
+	strcpy(plain_board, board + STREAM_LENGTH);
+	strcpy(tmp_board, plain_board);
+
+	if (!is_valid_stream(tmp_board))
+		return false;
+
+	// set difficulty level from save
+	int count = 0;
+	const char *tmp = plain_board;
+	while((tmp = strchr(tmp, '.')) != NULL)
+	{
+		count++;
+		tmp++;
+	}
+
+	g_resume_level = count;
+
+	return true;
+}
+
+bool save_stream(char user_board[], char plain_board[], int n)
+{
+	char* file_path = get_saved_file_path();
+	FILE* fp;
+	fp = fopen(file_path, "w");
+	if ( fp == NULL )
```

Comment:
Put the `free(file_path);` directly after the `fopen` because otherwise you might leak memory in the `return false` below :)